types/mempool/mempool_test.go

```diff
@@ -0,0 +1,785 @@
+
+func TestDefaultMempool(t *testing.T) {
+	ctx := sdk.NewContext(nil, tmproto.Header{}, false, log.NewNopLogger())
+	accounts := simtypes.RandomAccounts(rand.New(rand.NewSource(0)), 10)
+	txCount := 1000
+	var txs []testTx
+
+	for i := 0; i < txCount; i++ {
+		acc := accounts[i%len(accounts)]
+		tx := testTx{
+			address:  acc.Address,
+			priority: rand.Int63(),
+		}
+		txs = append(txs, tx)
+	}
+
+	// same sender-nonce just overwrites a tx
+	mp := mempool.NewPriorityMempool()
+	for _, tx := range txs {
+		ctx = ctx.WithPriority(tx.priority)
+		err := mp.Insert(ctx, tx)
+		require.NoError(t, err)
+	}
+	require.Equal(t, len(accounts), mp.CountTx())
+
+	// distinct sender-nonce should not overwrite a tx
+	mp = mempool.NewPriorityMempool()
+	for i, tx := range txs {
+		tx.nonce = uint64(i)
+		err := mp.Insert(ctx, tx)
+		require.NoError(t, err)
+	}
+	require.Equal(t, txCount, mp.CountTx())
+
+	sel, err := mp.Select(nil, 13)
+	require.NoError(t, err)
+	require.Equal(t, 13, len(sel))
+
+	// a tx which does not implement SigVerifiableTx should not be inserted
+	tx := &sigErrTx{getSigs: func() ([]txsigning.SignatureV2, error) {
+		return nil, fmt.Errorf("error")
+	}}
+	require.Error(t, mp.Insert(ctx, tx))
+	require.Error(t, mp.Remove(tx))
+	tx.getSigs = func() ([]txsigning.SignatureV2, error) {
+		return nil, nil
+	}
+	require.Error(t, mp.Insert(ctx, tx))
+	require.Error(t, mp.Remove(tx))
+
+	// removing a tx not in the mempool should error
+	mp = mempool.NewPriorityMempool()
+	require.NoError(t, mp.Insert(ctx, txs[0]))
+	require.ErrorIs(t, mp.Remove(txs[1]), mempool.ErrTxNotFound)
+	mempool.DebugPrintKeys(mp)
+
+	// inserting a tx with a different priority should overwrite the old tx
+	newPriorityTx := testTx{
+		address:  txs[0].address,
+		priority: txs[0].priority + 1,
+		nonce:    txs[0].nonce,
+	}
+	require.NoError(t, mp.Insert(ctx, newPriorityTx))
+	require.Equal(t, 1, mp.CountTx())
+}
+
+type txSpec struct {
+	i int
+	p int
+	n int
+	a sdk.AccAddress
+}
+
+func (tx txSpec) String() string {
+	return fmt.Sprintf("[tx i: %d, a: %s, p: %d, n: %d]", tx.i, tx.a, tx.p, tx.n)
+}
+
+func TestOutOfOrder(t *testing.T) {
+	accounts := simtypes.RandomAccounts(rand.New(rand.NewSource(0)), 2)
+	sa := accounts[0].Address
+	sb := accounts[1].Address
+
+	outOfOrders := [][]testTx{
+		{
+			{priority: 20, nonce: 1, address: sa},
+			{priority: 21, nonce: 4, address: sa},
+			{priority: 15, nonce: 1, address: sb},
+			{priority: 8, nonce: 3, address: sa},
+			{priority: 6, nonce: 2, address: sa},
+		},
+		{
+			{priority: 15, nonce: 1, address: sb},
+			{priority: 20, nonce: 1, address: sa},
+			{priority: 21, nonce: 4, address: sa},
+			{priority: 8, nonce: 3, address: sa},
+			{priority: 6, nonce: 2, address: sa},
+		}}
+
+	for _, outOfOrder := range outOfOrders {
+		var mtxs []mempool.Tx
+		for _, mtx := range outOfOrder {
+			mtxs = append(mtxs, mtx)
+		}
+		err := validateOrder(mtxs)
+		require.Error(t, err)
+	}
+
+	seed := time.Now().UnixNano()
```

Comment:
With this, if your unit test fails, how are you going to repro it? (At least you should log or print the seed)